Source/Core/Core/AchievementManager.cpp

```diff
@@ -55,614 +60,182 @@ void AchievementManager::SetUpdateCallback(UpdateCallback callback)
-                       name_to_fetch, m_display_name);
-          return;
-        }
-        m_player_badge.badge = std::move(fetched_badge);
-        m_player_badge.name = std::move(name_to_fetch);
-      }
-      else
-      {
-        WARN_LOG_FMT(ACHIEVEMENTS, "Failed to download player badge id {}.", name_to_fetch);
-      }
-
-      m_update_callback();
-    });
-  }
-
-  if (!IsGameLoaded())
-  {
-    m_update_callback();
-    return;
-  }
-
-  bool badgematch = false;
+  rc_client_achievement_list_t* achievement_list;
   {
     std::lock_guard lg{m_lock};
-    badgematch = m_game_badge.name == m_game_data.image_name;
+    achievement_list = rc_client_create_achievement_list(
+        m_client, RC_CLIENT_ACHIEVEMENT_CATEGORY_CORE_AND_UNOFFICIAL,
+        RC_CLIENT_ACHIEVEMENT_LIST_GROUPING_PROGRESS);
   }
-  if (!badgematch)
+  for (u32 bucket = 0; bucket < achievement_list->num_buckets; bucket++)
   {
-    m_image_queue.EmplaceItem([this] {
-      std::string name_to_fetch;
-      {
-        std::lock_guard lg{m_lock};
-        if (m_game_badge.name == m_game_data.image_name)
-          return;
-        name_to_fetch = m_game_data.image_name;
-      }
-      rc_api_fetch_image_request_t icon_request = {.image_name = name_to_fetch.c_str(),
-                                                   .image_type = RC_IMAGE_TYPE_GAME};
-      Badge fetched_badge;
-      if (RequestImage(icon_request, &fetched_badge) == ResponseType::SUCCESS)
-      {
-        INFO_LOG_FMT(ACHIEVEMENTS, "Successfully downloaded game badge id {}.", name_to_fetch);
-        std::lock_guard lg{m_lock};
-        if (name_to_fetch != m_game_data.image_name)
-        {
-          INFO_LOG_FMT(ACHIEVEMENTS, "Requested outdated badge id {} for game id {}.",
-                       name_to_fetch, m_game_data.image_name);
-          return;
-        }
-        m_game_badge.badge = std::move(fetched_badge);
-        m_game_badge.name = std::move(name_to_fetch);
-      }
-      else
-      {
-        WARN_LOG_FMT(ACHIEVEMENTS, "Failed to download game badge id {}.", name_to_fetch);
-      }
-
-      m_update_callback();
-    });
-  }
-
-  unsigned num_achievements = m_game_data.num_achievements;
-  for (size_t index = 0; index < num_achievements; index++)
-  {
-    std::lock_guard lg{m_lock};
-
-    // In case the number of achievements changes since the loop started; I just don't want
-    // to lock for the ENTIRE loop so instead I reclaim the lock each cycle
-    if (num_achievements != m_game_data.num_achievements)
-      break;
-
-    const auto& initial_achievement = m_game_data.achievements[index];
-    const std::string badge_name_to_fetch(initial_achievement.badge_name);
-    const UnlockStatus& unlock_status = m_unlock_map[initial_achievement.id];
-
-    if (unlock_status.unlocked_badge.name != badge_name_to_fetch)
+    for (u32 achievement = 0; achievement < achievement_list->buckets[bucket].num_achievements;
```

Comment:
To improve readability and avoid so many indexes, maybe save off the bucket:

```
const auto& bucket = achievement_list->buckets[bucket];
```

then

`achievement < bucket.num_achievements`

(as well as a number of places below)